### PR TITLE
ARROW-428: [Python] Multithreaded conversion from Arrow table to pandas.DataFrame

### PR DESCRIPTION
**Pull request by wesm**
This yields a substantial speedup on my laptop. On a 1GB numeric dataset, with 1 thread (the default prior to this patch):

```
>>> %timeit df2 = table.to_pandas(nthreads=1)
1 loop, best of 3: 498 ms per loop
```

With 4 threads (this is a true quad-core machine)

```
>>> %timeit df2 = table.to_pandas(nthreads=4)
1 loop, best of 3: 151 ms per loop
```

The default number of cores used is the `os.cpu_count` divided by 2 (since hyperthreading doesn't help with this largely memory-bound operation).

### CODE REVIEW

File: python/CMakeLists.txt [hunk 1]
```diff
@@ -66,7 +66,7 @@ endif(CCACHE_FOUND)
 ############################################################
 
 # compiler flags that are common across debug/release builds
-set(CXX_COMMON_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11 -Wall")
```
Comment (xhochy) [inline python/CMakeLists.txt]: That will make adding additional compiler flags via CMake harder?

File: python/pyarrow/table.pyx [hunk 2]
```diff
@@ -610,36 +612,29 @@ cdef class Table:
...
-            for i in range(self.table.num_columns()):
-                col = self.table.column(i)
-                column = self.column(i)
-                check_status(pyarrow.ConvertColumnToPandas(
-                    col, <PyObject*> column, &arr))
-                names.append(frombytes(col.get().name()))
-                data.append(PyObject_to_object(arr))
-
-            return pd.DataFrame(dict(zip(names, data)), columns=names)
+        if nthreads is None:
```
Comment (xhochy) [inline python/pyarrow/table.pyx]: As a good practice, I would also limit here to `environ['OMP_NUM_THREADS']`.

File: python/pyarrow/tests/test_convert_pandas.py [hunk 3]
```diff
@@ -27,6 +27,28 @@
...
+        'uint32': np.arange(size, dtype=np.uint32),
+        'uint64': np.arange(size, dtype=np.uint64),
+        'int8': np.arange(size, dtype=np.int16),
+        'int16': np.arange(size, dtype=np.int16),
+        'int32': np.arange(size, dtype=np.int32),
+        'int64': np.arange(size, dtype=np.int64),
+        'float32': np.arange(size, dtype=np.float32),
+        'float64': np.arange(size, dtype=np.float64),
+        'bool': np.random.randn(size) > 0,
+        # Pandas only support ns resolution, Arrow at the moment only ms
```
Comment (xhochy) [inline python/pyarrow/tests/test_convert_pandas.py]: Arrow also support s, us, ns. Just Parquet is limited to ms and us.

File: python/src/pyarrow/adapters/pandas.cc [hunk 4]
```diff
@@ -17,17 +17,18 @@
 
 // Functions for pandas conversion via NumPy
 
-#include <Python.h>
```
Comment (xhochy) [inline python/src/pyarrow/adapters/pandas.cc]: I think this include was needed for older Numpies, e.g. 1.8 and 1.9. (We should still support them as those are the default ones you should build `manylinux1` packages against.) Will re-add (with an explanative comment) if this was a problematic change.

Comment (wesm) [inline python/src/pyarrow/adapters/pandas.cc]: It's the first include in pyarrow/adapters/pandas.h, so this should have no effect

Comment (wesm) [inline python/src/pyarrow/adapters/pandas.cc]: I'll add it back in IWYU spirit

File: python/pyarrow/tests/test_convert_pandas.py (hunk 3 above)
Comment (wesm) [inline python/pyarrow/tests/test_convert_pandas.py]: Added a TODO. After the changes in this PR we should return to the timestamp resolution stuff in a separate patch

File: python/pyarrow/table.pyx (hunk 2 above)
Comment (wesm) [inline python/pyarrow/table.pyx]: done (it uses OMP_NUM_THREADS by default if it's set)

File: python/CMakeLists.txt (hunk 1 above)
Comment (wesm) [inline python/CMakeLists.txt]: Factored out common cmake code between C++/Python libraries, so you can set `CMAKE_CXX_FLAGS` to modify externally

**Comment by wesm** [comment]
+1

**Pull request closed without merging**